File: app/src/main/java/com/example/walmartpathfinder/models/Instantiator.kt
package com.example.walmartpathfinder.models

class Instantiator {
    companion object {
        var mainSections = mutableListOf<Section>()
        var subSections = mutableListOf<Section>()
        var items = mutableListOf<Item>()

        /* Store structure with sections and subsections*/
        fun createStoreStructure(): Store {
            val store = Store("Walmart")

            val mainSecNames = listOf<String>("Dairy", "Produce", "Meat", "Frozen", "Bakery")
            val subSecNames = listOf<Collection<String>>(
                listOf("A1", "A2", "A3", "A4"),
                listOf("B1", "B2", "B3"),
                listOf("C1", "C2", "C3", "C4"),
                listOf("D1", "D2"),
                listOf()
            )



            /* Add all section names and subsections names to store */
            for ((index, mainSecName) in mainSecNames.withIndex()) {
                val mainSec = ConcreteSection(mainSecName)
                store.section.addSubsection(mainSec)
                mainSections.add(mainSec)

                for (subSecName in subSecNames[index]) {
                    val subSec = ConcreteSection(subSecName)
                    mainSec.addSubsection(subSec)
                    subSections.add(subSec)
                }
            }

            val dairy = store.section.getSubsections()[0] as ConcreteSection
            (dairy.getSubsections()[0] as ConcreteSection).addSubsection(Item("Milk",dairy.getSubsections()[0]))
            (dairy.getSubsections()[1] as ConcreteSection).addSubsection(Item("Buttermilk",dairy.getSubsections()[1]))
            (dairy.getSubsections()[2] as ConcreteSection).addSubsection(Item("Yogurt",dairy.getSubsections()[2]))
            (dairy.getSubsections()[3] as ConcreteSection).addSubsection(Item("Butter",dairy.getSubsections()[3]))


            val produce = store.section.getSubsections()[1] as ConcreteSection
            (produce.getSubsections()[0] as ConcreteSection).addSubsection(Item("Cucumbers",produce.getSubsections()[0]))
            (produce.getSubsections()[1] as ConcreteSection).addSubsection(Item("Onions",produce.getSubsections()[1]))
            (produce.getSubsections()[2] as ConcreteSection).addSubsection(Item("Tomatoes",produce.getSubsections()[2]))

            val meat = store.section.getSubsections()[2] as ConcreteSection
            (meat.getSubsections()[0] as ConcreteSection).addSubsection(Item("Pork", meat.getSubsections()[0]))
            (meat.getSubsections()[1] as ConcreteSection).addSubsection(Item("Beef", meat.getSubsections()[1]))
            (meat.getSubsections()[2] as ConcreteSection).addSubsection(Item("Chicken", meat.getSubsections()[2]))
            (meat.getSubsections()[4] as ConcreteSection).addSubsection(Item("Ground meat", meat.getSubsections()[4]))

            val frozen = store.section.getSubsections()[3] as ConcreteSection
            (frozen.getSubsections()[0] as ConcreteSection).addSubsection(Item("Ice Cream", frozen.getSubsections()[0]))
            (frozen.getSubsections()[1] as ConcreteSection).addSubsection(Item("Waffles", frozen.getSubsections()[1]))
            (frozen.getSubsections()[1] as ConcreteSection).addSubsection(Item("Pizza", frozen.getSubsections()[1]))

            val bakery = store.section.getSubsections()[4] as ConcreteSection
            bakery.addSubsection(Item("Bread", bakery))
            bakery.addSubsection(Item("Cake", bakery))
            bakery.addSubsection(Item("Muffins",bakery))

            return store
        }

        @JvmStatic
        fun main(args: Array<String>) {
            var walmart = createStoreStructure()
            println(mainSections)
            println(subSections)
            println(items)

            TODO("demo use of generateList() on an item list")
        }
    }
}

File: src/models/Instantiator.ts
import { ConcreteSection } from "./ConcreteSection";
import { Item } from "./Item";
import { Section } from "./Section";
import { Store} from "./Store";

class Instantiator{
    static mainSections: Section[] = []
    static subSections: Section[] = []
    static store: Store | undefined = undefined
    static items: Item[] = []
    
    static createItemInSubsection(section: ConcreteSection, subSectionIdx: number, itemName: string){
        const item = new Item(itemName, section.getSubsections()[subSectionIdx]);
        (section.getSubsections()[subSectionIdx] as ConcreteSection).addSubsection(item)
        this.items.push(item)
    }

    static createItem(section: ConcreteSection, itemName: string){
        const item = new Item(itemName, section)
        section.addSubsection(item)
        this.items.push(item)
    }

    static createStoreStructure(){
        const store = new Store("Walmart");

        const mainSectionNames: string[] = ["Dairy", "Produce", "Meat", "Frozen", "Bakery"] 
        const subSectionNames = [
            ["A1", "A2", "A3", "A4"],
            ["B1", "B2", "B3"],
            ["C1", "C2", "C3", "C4"],
            ["D1", "D2"],
            []
        ]

        for (let i = 0; i < mainSectionNames.length; i ++){
            let mainSec = new ConcreteSection(mainSectionNames[i], undefined);
            store.section.addSubsection(mainSec);
            this.mainSections.push(mainSec);

            for (let j = 0; j < subSectionNames[i].length; j ++){
                let subSec = new ConcreteSection(subSectionNames[i][j], undefined)
                mainSec.addSubsection(subSec)
                this.subSections.push(subSec)
            }
        }

        let dairy = store.section.getSubsections()[0] as ConcreteSection
        this.createItemInSubsection(dairy, 0, "Milk")
        this.createItemInSubsection(dairy, 1, "Buttermilk")
        this.createItemInSubsection(dairy, 2, "Yogurt")
        this.createItemInSubsection(dairy, 2, "Kefir")
        this.createItemInSubsection(dairy, 3, "Butter")

        let produce = store.section.getSubsections()[1] as ConcreteSection
        this.createItemInSubsection(produce, 0, "Cucumbers")
        this.createItemInSubsection(produce, 1, "Onions")
        this.createItemInSubsection(produce, 2, "Tomatoes")

        let meat = store.section.getSubsections()[2] as ConcreteSection
        this.createItemInSubsection(meat, 0, "Pork")
        this.createItemInSubsection(meat, 1, "Beef")
        this.createItemInSubsection(meat, 2, "Chicken")
        this.createItemInSubsection(meat, 3, "Ground meat")

        /* Items in the Frozen Section */
        let frozen = store.section.getSubsections()[3] as ConcreteSection
        this.createItemInSubsection(frozen, 0, "Ice Cream")
        this.createItemInSubsection(frozen, 1, "Waffles")
        this.createItemInSubsection(frozen, 1, "Pizza")

        /* Items in the Bakery Section */
        let bakery = store.section.getSubsections()[4] as ConcreteSection
        this.createItem(bakery, "Bread")
        this.createItem(bakery, "Cake")
        this.createItem(bakery, "Muffins")

        this.store = store;
    }
}

export {Instantiator}